### PR TITLE
Add Kokkos smoke test support

### PR DESCRIPTION
**Pull request by janciesko**
Adds support for Kokkos Spack test.

### CODE REVIEW

**Comment by alalazo** [comment]
@jjwilke

**Comment by crtrott** [comment]
@jjwilke is a manager now not sure he still actively monitors stuff like this.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (jjwilke) [inline var/spack/repos/builtin/packages/kokkos/package.py]: you can list both myself (jjwilke) and jciesko

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (jjwilke) [inline var/spack/repos/builtin/packages/kokkos/package.py:23]: I will trust this is the correct sha!

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (jjwilke) [inline var/spack/repos/builtin/packages/kokkos/package.py:310]: What is the working directory here? This might clash with build directory.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (jjwilke) [inline var/spack/repos/builtin/packages/kokkos/package.py:294]: I would do this test via version number. I forget exactly, though, how to test for master/devel.  If someone breaks this by moving directories around, this will silently not run when we would want it to crash.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (jjwilke) [inline var/spack/repos/builtin/packages/kokkos/package.py:289]: Can we document approximately here what order/when we expect these functions to get invoked?

**Review by jjwilke — CHANGES_REQUESTED** [review]
Looks good. A couple nitpicky things.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/kokkos/package.py:328]:
```suggestion
        if not os.path.exists(cmake_path):
            print('Skipping smoke tests: {0} is missing'.format(cmake_path))
```

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/kokkos/package.py:328]: As written, the test log has no information on the smoke test when the script is missing other than indicating the test passed, giving the impression everything went fine even thought the smoke test was skipped.  For example, I'm getting the following:

```
==> Results for test suite 'kokkos':
==>   kokkos-3.3.01-xjr6jkq PASSED
==> Testing package kokkos-3.3.01-xjr6jkq
```

Even when the smoke test is skipped.

While this may be desirable in terms of not claiming failure if the smoke test cannot run, it can be misleading; hence, the suggestion of noting when the smoke test is skipped.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/kokkos/package.py:310]: Chiming in here: As long as this is called from the context of Spack smoke tests (i.e., from within `def test`), the directory will be within the package's (smoke) test stage directory.  (See https://spack.readthedocs.io/en/latest/packaging_guide.html?highlight=test_stage#install-tests.)

Comment (janciesko) [reply var/spack/repos/builtin/packages/kokkos/package.py:310]: Yes, we should be good here.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (janciesko) [reply var/spack/repos/builtin/packages/kokkos/package.py:289]: I would propose omitting documentation as people who'll be reading this probably know that. ```"@run_after('install')"``` runs after the install step where we move test source files to the .spack/test dir. During ```spack test run```, we build and run.

**Comment by alalazo** [comment]
> @jjwilke is a manager now not sure he still actively monitors stuff like this.

@crtrott I've been pinging @jjwilke as the maintainer of the package. It's absolutely fine if circumstances changed and people don't have time anymore for that (indeed we are grateful for the work done so far!). This is just to say that you should feel free to remove people from the list if they don't want to be pinged anymore or add other maintainers too if they want to share the burden  of reviews.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (janciesko) [reply var/spack/repos/builtin/packages/kokkos/package.py:294]: @tldahlgren - Can we check for version during test build?

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/kokkos/package.py:294]: Yes, you should definitely be able to check for the version.  Let me know if you cannot.

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/kokkos/package.py:294]: We generally recommend using `self.spec.satisfies('@[<start>]:[<end>]')` for this purpose, where you can specify the starting version `<start>` and or the ending version `<end>` that applies since it can be used to perform version range checks as well as other constraints on the specification.

**Comment by tldahlgren** [comment]
@janciesko Did you want to add a version check to `setup_build_tests` before the PR is merged?

**Comment by tldahlgren** [comment]
FWIW. I have built and run the software at the default (where I get the test skip message) and at develop (where the tests run and pass).

**Comment by janciesko** [comment]
> @janciesko Did you want to add a version check to `setup_build_tests` before the PR is merged?

We can merge asap so folks can test and improve version check soon.

**Comment by tldahlgren** [comment]
Please resolve the style check:

```
var/spack/repos/builtin/packages/kokkos/package.py:19: [E231] missing whitespace after ','
```

**Review by tldahlgren — APPROVED** [review]
Approving contingent on all required tests passing and assuming the requested version check will be addressed in a subsequent PR.

**Comment by krcb** [comment]
@tldahlgren @janciesko There appears to be an issue in the code that was added here. Specifically, I am seeing:

```
==> kokkos: Executing phase: 'cmake'
==> kokkos: Executing phase: 'build'
==> kokkos: Executing phase: 'install'
==> Error: NameError: name 'cmake' is not defined

$SPACK_ROOT/spack/var/spack/repos/builtin/packages/kokkos/package.py:351, in setup_build_tests:
        348            ),
        349            "-DSPACK_PACKAGE_INSTALL_DIR:PATH={0}".format(self.prefix),
        350        ]
  >>    351        cmake(*cmake_args)
        352        self.cache_extra_test_sources(cmake_out_path)
        353        self.cmake_bin(set=True)
```

I'm not sure how this is not being caught in other places, but I think something like the following needs to be added to the Kokkos package file to ensure that the cmake binary is defined:

```
cmake = Executable(self.spec["cmake"].prefix.bin.cmake)
```

**Comment by tldahlgren** [comment]
> ```
> cmake = Executable(self.spec["cmake"].prefix.bin.cmake)
> ```

I am seeing that too.  I thought we were making that available at one point but it clearly isn't now. 

You can also use `cmake = which(self.spec["cmake"].prefix.bin.cmake))`

**Comment by krcb** [comment]
@tldahlgren Shall I open an issue to address this? This is preventing us from using spack in a project that uses Kokkos right now.

**Comment by tldahlgren** [comment]
> @tldahlgren Shall I open an issue to address this? This is preventing us from using spack in a project that uses Kokkos right now.

Up to you.  I'm addressing it as it relates to stand-alone tests in #34236 and its off-shoots.

**Pull request merged**
